### ISSUE TITLE
Popover Component Feature Request

### ISSUE THREAD

**Issue by brmendez** [sir-dunxalot/ember-tooltips#77]
Hello and thanks for ember-tooltips. Any plans to bring in a Popover component? 

Thank you.

**Comment by sir-dunxalot** [comment]
Could do! It would be pretty simple to execute right now with a simple class and some styling. Are there any functionality differences you would expect between a tooltip and a popover?

**Comment by brmendez** [comment]
Hi sir-dunxalot, thanks for the fast reply!

Pretty much the same functionality as Bootstraps Popover component (http://v4-alpha.getbootstrap.com/components/popovers/). I was originally on the search for Bootstraps Popover and Tooltip component for Ember, but came across your Tooltip, which works great! No luck on Popover's though. Most of the options currently available (that I have seen) are either outdated or broken.

The key component I am looking for is the ability to have a **header/title** as a part of the Popover (http://v4-alpha.getbootstrap.com/components/popovers/#live-demo) or (http://www.tutorialrepublic.com/lib/images/bootstrap/twitter-bootstrap-popover.png). With the options/attributes you already have available in your Tooltip - E.g. tethering it to any one of the four corners of an element, the ability to have static Popovers, as well as having a programatic way to hide and show a Popover. 

I hope this answers your question. Thank you sir-duxalot (Duncan)! Please let me know if I can provide any additional clarity.

Brian

**Comment by sir-dunxalot** [comment]
Ok, sounds good! I think this would be a good addition. I'll have a think about how best to implement this but let me know if you have any ideas.

**Comment by andrew-zenefits** [comment]
{x-post from https://github.com/sir-dunxalot/ember-tooltips/issues/51#issuecomment-227873282}

I agree with @brmendez that a popover would be a great functionality to have in ember-tooltips. Here is how we at Zenefits want to differentiate between tooltips and popovers...
* tooltip
  * not interactive
  * used for showing small amounts of helper text (we're considering a 140 character limit at Zenefits)
  * black background, white text 
  * this library already covers that 100%
* popover
  * interactive: interior can include links and other interactive components (radio buttons, select, youTube videos, etc)
  * used for showing interactive elements (see pic)
  * different UI than a tooltip to signify that interaction is possible. This could be standardized by this library or by developers extending a base popover component and adding their own handling for title, body, yields, and styles (I prefer the second approach for customization)

I've been playing around with different ways of achieving popover functionality and have been unable to get it to work. @sir-dunxalot I have some free cycles and would be happy to build a solution but would need some guidance on the best way to do this. 


<img width="750" alt="screen shot 2016-06-22 at 1 38 56 pm" src="https://cloud.githubusercontent.com/assets/8782940/16282697/bddd3d92-387e-11e6-878f-d3793746d232.png">

----------------NEW COMMENTS------------

@sir-dunxalot I would be happy to digging into the addon and figuring out a way to implement it. Should we keep all discussion in this thread? If so I can start commenting with what I've tried and the ways those attempts have fallen short. Unfortunately I don't have a great idea of the correct approach but would be happy to make a PR if I can be pointed in the right implementation direction.

**Comment by sir-dunxalot** [comment]
Yeah, good idea - let's keep the discussion in this thread.

As for an implementation, I can immediately think of three options:

1) Extend the current component like `{{tooltip-on-element}}` as `{{popover-on-element}}`
2) Add a `isPopover` property to the current components
3) Refactor the current tooltip components to share some base code between `{{tooltip-on-element}}` and `{{popover-on-element}}`

I would learn towards 1 or 3 myself - what do you think?

I doubt there would need to be too much extra functionality added outside of event handling - we will be able to have a `.popover` class used instead of a `.tooltip` class for styling. I think we should try and support as many currently supported `showOn` events as possible even if the close only happens when clicking outside the popover or maybe clicking a little close `x`.

One other decision that will have to be made is how much we want to structure content of popovers vs requiring developers to do it themselves. For example, do we have a `{{popover-header}}` component? Do we have a `title` property on popovers and have the block content shown in the `{{yield}}`? Etc.

**Comment by MiguelMadero** [comment]
I would lean more towards 3 or 1. I think it makes more sense not to do anything for layout, since apps will always have different concerns and it's easy to do this internally on the block or provide a common abstraction for an app as needed. For example: 

```
{{#popover-on-element}}
   {{#my-custom-popover-layout as layout}}
      {{#layout.header}}Some text{{/layout.header}}
      {{#layout.body}}lorem ipsum{{#layout.body}}
   {{/my-custom-popover-layout}}
{{/popover-on-element}}
```

**Comment by andrew-zenefits** [comment]
Thanks for your thoughts guys. I'll hopefully get started on this soon.

**Comment by sir-dunxalot** [comment]
On the contrary - thank you. Keep firing questions and ideas our way!

**Comment by andrew-zenefits** [comment]
I have a working prototype here https://github.com/zenefits/ember-tooltips/pull/3

@sir-dunxalot, @MiguelMadero could you both review and let me know what your thoughts are on (1) code structure and (2) popover UI

When we finalize the PR I can either dogfood it Zenefits for a bit then merge it into ember-tooltips, or make a PR directly to this library first. 

Thanks!

**Comment by MiguelMadero** [comment]
@andrew-zenefits the change looks good to me, but I would like to make sure we don't deviate from the original one and to try our best to keep a single codebase. Is the change backwards compatible? 

@sir-dunxalot what do you think about zenefits#3?

**Comment by sir-dunxalot** [comment]
I'm out tomorrow but I will take a look later this weekend or at the start of next week. Sorry for the delay. On first glance this looks like a great step forwards.

**Comment by andrew-zenefits** [comment]
@MiguelMadero & @sir-dunxalot here is the newest PR. I redid it in a better fashion and it made sense to just start from the ground up.

https://github.com/zenefits/ember-tooltips/pull/8

**Comment by sir-dunxalot** [comment]
Hi gents,

I haven't forgotten about this - still trying to get to it. If this is delaying anything at Zenefits, you could go ahead and merge your PR and issue a PR directly to this library. I should be able to review this by the end of tomorrow (10th).

**Comment by jamesdixon** [comment]
@sir-dunxalot any update on this?

**Comment by jamesdixon** [comment]
Nevermind -- just saw the PR referenced here 👍

**Comment by sir-dunxalot** [comment]
Fixed by #92.
